Life-Pill/phamacy-pos-employee-service
# Add Employer Communication and Unit Tests

# Pull Request: Add Employer Communication and Unit Tests

## Title
Add new endpoints and unit tests for EmployerController

## Description

This pull request includes several new endpoints and updates to the `EmployerController` class to handle various operations related to employer management. The changes ensure that employer-related data can be saved, retrieved, updated, and deleted as required by the application. Additionally, unit tests have been implemented to validate the functionality of these endpoints.

### Changes Made

#### Endpoints Added:
- `POST /save-without-image`: Save an employer without an image.
- `POST /save-with-image`: Save an employer with an image.
- `GET /profile-photo/{employerId}`: Retrieve the profile photo of an employer by ID.
- `PUT /update/{employerId}`: Update details of an employer.
- `PUT /updateAccountDetails`: Update account details of an employer.
- `PUT /updateEmployerBankAccountDetailsWithId/{employerId}`: Update bank account details of an employer.
- `GET /get-by-employee-id`: Retrieve an employer by ID.
- `GET /get-all-employee-details-by-id`: Retrieve all details of an employer by ID.
- `GET /view-image/{employerId}`: Retrieve the profile photo of an employer by ID.
- `DELETE /delete-employer-by-Id/{employerId}`: Delete an employer by ID.
- `GET /get-all-employers`: Retrieve all employers.
- `GET /get-all-employers-by-active-state/{status}`: Retrieve all employers by active state.
- `GET /get-all-employers-bank-details`: Retrieve bank details of all employers.
- `GET /bank-details/{employerId}`: Retrieve bank details of an employer by ID.

### Test Coverage:
- Implemented unit tests to cover various scenarios for the added endpoints.
- Ensured test cases verify the correct behavior and responses for successful operations and edge cases.

## Affected Files

- `EmployerController.java`
- `EmployerService.java` (assumed updates for service methods)
- `EmployerMapper.java` (assumed updates for mapping methods)
- `EmployerControllerTest.java` (new unit tests)

## Test Plan

- **Unit Tests**: Run the implemented unit tests for `EmployerController` using JUnit to ensure all methods function as expected.
- **Manual Testing**: Perform manual testing of each endpoint in a local environment to validate correct functionality and responses.

### Test Scenarios:
1. Save an employer without an image and verify a successful response.
2. Save an employer with an image and verify the image is correctly saved.
3. Retrieve an employer's profile photo by ID and check the returned image.
4. Update an employer's details and verify the updated information.
5. Update an employer's account details and check for a successful update.
6. Update an employer's bank account details and ensure the bank information is updated.
7. Retrieve an employer by ID and validate the returned employer details.
8. Retrieve all details of an employer by ID and verify all returned information.
9. Delete an employer by ID and ensure the employer is removed.
10. Retrieve all employers and validate the list of returned employers.
11. Retrieve all employers by active state and ensure the returned list matches the state.
12. Retrieve bank details of all employers and verify the returned data.
13. Retrieve bank details of an employer by ID and check the returned bank details.

## Related Issues

- [Issue #1  ]: Employee service Application

## Checklist

- [x] Added endpoints for employer management operations.
- [x] Implemented unit tests for all new endpoints.
- [x] Manually tested endpoints for correct functionality.

## Conclusion

This pull request enhances the `EmployerController` class by adding new endpoints and corresponding unit tests to manage employer-related operations effectively. Please review the changes and provide feedback or suggestions for further improvement.

Best regards,
Pramitha Jayasooriya,
Backend Developer at LifePill,
https://pramithamj.me

---
## Donation

***If you like what I do, maybe consider buying me a coffee***

<a href="https://buymeacoffee.com/lpramithamm"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-red.png" alt="Buy Me A Coffee" style="height: 35px !important; width: 120px !important;"></a>

***
